client/tx/tx.go

```diff
@@ -36,3 +50,301 @@ type (
+			return
+		}
+
+		_, adjusted, err := CalculateGas(ctx.QueryWithData, txf, msgs...)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		txf = txf.WithGas(adjusted)
+
+		if br.Simulate {
+			rest.WriteSimulationResponse(w, ctx.Marshaler, txf.Gas())
+			return
+		}
+	}
+
+	tx, err := BuildUnsignedTx(txf, msgs...)
+	if err != nil {
+		rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	output, err := ctx.Marshaler.MarshalJSON(tx)
+	if err != nil {
+		rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(output)
+}
+
+// BuildUnsignedTx builds a transaction to be signed given a set of messages. The
+// transaction is initially created via the provided factory's generator. Once
+// created, the fee, memo, and messages are set.
+func BuildUnsignedTx(txf Factory, msgs ...sdk.Msg) (ClientTx, error) {
+	if txf.chainID == "" {
+		return nil, fmt.Errorf("chain ID required but not specified")
+	}
+
+	fees := txf.fees
+	if !txf.gasPrices.IsZero() {
+		if !fees.IsZero() {
+			return nil, errors.New("cannot provide both fees and gas prices")
+		}
+
+		glDec := sdk.NewDec(int64(txf.gas))
+
+		// Derive the fees based on the provided gas prices, where
+		// fee = ceil(gasPrice * gasLimit).
+		fees = make(sdk.Coins, len(txf.gasPrices))
+		for i, gp := range txf.gasPrices {
+			fee := gp.Amount.Mul(glDec)
+			fees[i] = sdk.NewCoin(gp.Denom, fee.Ceil().RoundInt())
+		}
+	}
+
+	tx := txf.txGenerator.NewTx()
+	tx.SetFee(auth.NewStdFee(txf.gas, fees))
+	tx.SetMsgs(msgs...)
+	tx.SetMemo(txf.memo)
+	tx.SetSignatures()
+
+	return tx, nil
+}
+
+// BuildSimTx creates an unsigned tx with an empty single signature and returns
+// the encoded transaction or an error if the unsigned transaction cannot be
+// built.
+func BuildSimTx(txf Factory, msgs ...sdk.Msg) ([]byte, error) {
+	tx, err := BuildUnsignedTx(txf, msgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an empty signature literal as the ante handler will populate with a
+	// sentinel pubkey.
+	tx.SetSignatures(auth.NewStdSignature(nil, nil))
+
+	return tx.Marshal()
+}
+
+// CalculateGas simulates the execution of a transaction and returns the
+// simulation response obtained by the query and the adjusted gas amount.
+func CalculateGas(
+	queryFunc func(string, []byte) ([]byte, int64, error), txf Factory, msgs ...sdk.Msg,
+) (sdk.SimulationResponse, uint64, error) {
+
+	txBytes, err := BuildSimTx(txf, msgs...)
+	if err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
+
+	bz, _, err := queryFunc("/app/simulate", txBytes)
+	if err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
+
+	var simRes sdk.SimulationResponse
+	if err := json.Unmarshal(bz, &simRes); err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
```

Comment:
imho there should be a util function `Simulate` that wraps this code

Reply:
I don't follow. `CalculateGas` is the util function.